Convert misc chapters to prodn, update syntax (#12936)
coq/coq: doc/sphinx/user-extensions/proof-schemes.rst

Generation of induction principles with ``Scheme``
--------------------------------------------------------

.. cmd:: Scheme {? @ident := } @scheme_kind {* with {? @ident := } @scheme_kind }

   .. insertprodn scheme_kind sort_family

   .. prodn::
      scheme_kind ::= Equality for @reference
      | {| Induction | Minimality | Elimination | Case } for @reference Sort @sort_family
      sort_family ::= Set
      | Prop
      | SProp
      | Type

  A high-level tool for automatically generating
  (possibly mutual) induction principles for given types and sorts.

>   Each :n:`@ident` is a different inductive type identifier belonging to

Reviewer: That's `@reference` here. It seems that you have marked this as resolved without fixing it.

Author: It's so hard to find good workers these days, especially with the low pay.  :-)

  the same package of mutual inductive definitions.
  The command generates the :n:`@ident`\s as mutually recursive
  definitions. Each term :n:`@ident` proves a general principle of mutual
  induction for objects in type :n:`@reference`.

>   :n:`@reference`

Reviewer: That's `@ident` here.

    The name of the scheme. If not provided, the scheme name will be determined automatically
    from the sorts involved.

>   :n:`Minimality`

Reviewer: For consistency with the presentation of the `Equality` clause below, this should be:
```suggestion
  :n:`Minimality for @reference Sort @sort_family`
```

(possibly mutual) induction principles for given types and sorts. Its

> syntax follows the schema:

Reviewer: This sentence is repeated in the description of the command below, so I would simply remove it.

> .. cmd:: Scheme {? @ident__def := } @scheme_kind {* with {? @ident__def := } @scheme_kind }

Reviewer: Since there is no longer any `ident` besides the `@ident__def`, I don't think keeping the `__def` is that useful. I would remove it.

   .. old Scheme @ident__1 := Induction for @ident__2 Sort @sort {* with @ident__i := Induction for @ident__j Sort @sort}

   .. insertprodn scheme_kind scheme_kind

   .. prodn::
      scheme_kind ::= Equality for @reference
      | {| Induction | Minimality | Elimination | Case } for @reference Sort @sort_family

  A high-level tool for automatically generating
  (possibly mutual) induction principles for given types and sorts.

>   Each :n:`@ident__def` is a different inductive type identifier belonging to

Reviewer: ```suggestion
  Each :n:`@reference` is a different inductive type identifier belonging to
```

Combined Scheme
~~~~~~~~~~~~~~~~~~~~~~

> .. cmd:: Combined Scheme @ident from {+, @ident }

Reviewer: OTOH, in this case, removing the indices creates ambiguity, so if you don't like the `__i`, please introduce a `__def` for the first `@ident`.

   This command is a tool for combining induction principles generated
   by the :cmd:`Scheme` command.
   Each :n:`@ident` is a different inductive principle that must  belong
   to the same package of mutual inductive principle definitions.

>    This command generates :n:`@ident` as the conjunction of the

Reviewer: This will become `@ident_def`.

Generation of inversion principles with ``Derive`` ``Inversion``
-----------------------------------------------------------------

.. cmd:: Derive Inversion @ident with @one_term {? Sort @sort_family }

   Generates the inversion lemma for the
   :tacn:`inversion ... using ...` tactic.  :token:`ident` is the name

>    of the generated principle.  :token:`one_term` should be an inductive

Author: Take a look at what I wrote.

>    of the generated principle.  :token:`one_term` should be an inductive

Author: Please check what I wrote.

Generation of inversion principles with ``Derive`` ``Inversion``
-----------------------------------------------------------------

.. cmd:: Derive Inversion @ident with @one_term {? Sort @sort_family }

   Generates an inversion lemma for the
   :tacn:`inversion ... using ...` tactic.  :token:`ident` is the name
   of the generated lemma.  :token:`one_term` should be in the form
   :token:`qualid` or :n:`(forall {+ @binder }, @qualid @term)` where
   :token:`qualid` is the name of an inductive principle
   an inductive

>    predicate, and :n:`{* @binder }` has the variables occurring in the term

Reviewer: What about:
```suggestion
   predicate, and :n:`{+ @binder }` captures the variables occurring in the term
```
?

Author: How's this?

```
where
   :token:`qualid` is the name of an inductive 
   predicate and :n:`{+ @binder }` binds the variables occurring in the term
   :token:`term`.
```

Reviewer: Yes, sounds good.

Generation of inversion principles with ``Derive`` ``Inversion``
-----------------------------------------------------------------

.. cmd:: Derive Inversion @ident with @one_term {? Sort @sort_family }

   .. insertprodn sort_family sort_family

   .. prodn::
      sort_family ::= Set
      | Prop
      | SProp

>       | Type

Reviewer: I don't get why `sort_family` is inserted so late when it's actually also used in `Scheme`.

   This command generates an inversion principle for the

>    :tacn:`inversion ... using ...` tactic.  The first :token:`ident` is the name

Reviewer: This description is now incorrect. There should be a description of what `one_term` can be.
